File: tools/scan_used_stridx_bidx.py
#!/usr/bin/env python2
#
#  Scan Duktape code base for references to built-in strings and built-in
#  objects, i.e. for:
#
#  - Strings which will need DUK_STRIDX_xxx constants and a place in the
#    thr->strs[] array.
#
#  - Objects which will need DUK_BIDX_xxx constants and a place in the
#    thr->builtins[] array.
#

import os
import sys
import re
import json

re_str_stridx = re.compile(r'DUK_STRIDX_(\w+)', re.MULTILINE)
re_str_heap = re.compile(r'DUK_HEAP_STRING_(\w+)', re.MULTILINE)
re_str_hthread = re.compile(r'DUK_HTHREAD_STRING_(\w+)', re.MULTILINE)
re_obj_bidx = re.compile(r'DUK_BIDX_(\w+)', re.MULTILINE)

def main():
    str_defs = {}
    obj_defs = {}

    for fn in sys.argv[1:]:
        with open(fn, 'rb') as f:
            d = f.read()
            for m in re.finditer(re_str_stridx, d):
                str_defs[m.group(1)] = True
            for m in re.finditer(re_str_heap, d):
                str_defs[m.group(1)] = True
            for m in re.finditer(re_str_hthread, d):
                str_defs[m.group(1)] = True
            for m in re.finditer(re_obj_bidx, d):
                obj_defs[m.group(1)] = True

    str_used = []
    for k in sorted(str_defs.keys()):
        str_used.append('DUK_STRIDX_' + k)

    obj_used = []
    for k in sorted(obj_defs.keys()):
        obj_used.append('DUK_BIDX_' + k)

    doc = {
        'used_stridx_defines': str_used,
        'used_bidx_defines': obj_used,
        'count_used_stridx_defines': len(str_used),
        'count_used_bidx_defines': len(obj_used)
    }
    print(json.dumps(doc, indent=4))

if __name__ == '__main__':
    main()

File: src-tools/lib/configure/scan_used_stridx_bidx.js
'use strict';

const { readFileUtf8 } = require('../util/fs');

// Scan source files for:
// - strings which need a stridx (to thr->strs[])
// - objects which need a bidx (to thr->builtins[])
// - DUK_USE_xxx config options referenced
function scanUsedStridxBidx(filenameList) {
    var strDefs = {};
    var objDefs = {};
    var optDefs = {};

    function matchAll(data, re, map) {
        var m;
        while ((m = re.exec(data)) !== null) {
            map[m[1]] = true;
        }
    }

    for (let fn of filenameList) {
        const re_str_stridx = /DUK_STRIDX_(\w+)/gm;
        const re_str_heap = /DUK_HEAP_STRING_(\w+)/gm;
        const re_str_hthread = /DUK_HTHREAD_STRING_(\w+)/gm;
        const re_obj_bidx = /DUK_BIDX_(\w+)/gm;
        const re_duk_use = /DUK_USE_(\w+)/gm;

        let data = readFileUtf8(fn);
        matchAll(data, re_str_stridx, strDefs);
        matchAll(data, re_str_heap, strDefs);
        matchAll(data, re_str_hthread, strDefs);
        matchAll(data, re_obj_bidx, objDefs);
        matchAll(data, re_duk_use, optDefs);
    }

    var strUsed = Object.keys(strDefs).sort().map((v) => 'DUK_STRIDX_' + v);
    var objUsed = Object.keys(objDefs).sort().map((v) => 'DUK_BIDX_' + v);
    var optUsed = Object.keys(optDefs).sort().map((v) => 'DUK_USE_' + v);

    return {
        usedStridxDefines: strUsed,
        usedBidxDefines: objUsed,
        usedDukUseOptions: optUsed,
    };
}
exports.scanUsedStridxBidx = scanUsedStridxBidx;
